src/hotspot/os/bsd/os_perf_bsd.cpp

```diff
@@ -34,7 +34,9 @@
   #include <mach/mach.h>
   #include <mach/task_info.h>
 #else
+#if !defined(__arm__)
```

Comment:
We have this header, the problem is the duplicate definition of the "frame" variable. We define it on armv6/7 in [frame.h](https://svnweb.freebsd.org/base/head/sys/arm/include/frame.h?revision=326258&view=markup#l133) and "frame" is also used in [frame.hpp](https://github.com/battleblow/openjdk-jdk11u/blob/master/src/hotspot/share/runtime/frame.hpp#L50)

This is the error message:
```
In file included from src/hotspot/os/bsd/os_perf_bsd.cpp:37:
In file included from /usr/include/sys/user.h:38:
In file included from /usr/include/machine/pcb.h:41:
/usr/include/machine/frame.h:131:8: error: redefinition of 'frame'
struct frame {
       ^
/usr/ports/java/openjdk11/work/openjdk-jdk11u-jdk-11.0.4-11-2/src/hotspot/share/runtime/frame.hpp:55:7: note: previous definition is here
class frame {
```

Reply:
Hmm, arm must be unusual here since this doesn't happen on other architectures.  I don't see frame.hpp explicitly in os_perf_bsd.cpp, so I wonder what is pulling it in?  We may be able to avoid one or the other definitions since neither are explicitly used in the code.

Reply:
I just put an `#error "mik frame.hpp" `in frame.hpp and attempted a build on amd64:
```
In file included from src/hotspot/share/precompiled/precompiled.hpp:37:         
In file included from src/hotspot/share/gc/shared/collectedHeap.hpp:36:         
In file included from src/hotspot/share/utilities/events.hpp:30:                
In file included from src/hotspot/share/runtime/thread.hpp:29:                  
In file included from src/hotspot/share/code/compiledMethod.hpp:28:             
In file included from src/hotspot/share/code/codeBlob.hpp:31:                   
src/hotspot/share/runtime/frame.hpp:37:2: error: "mik frame.hpp"                
#error "mik frame.hpp"
```

Reply:
My first suggestion then is to turn off pre-compiled headers on arm6/7 since I don't think you will pull in frame.hpp in that case.

Reply:
It changes nothing, codeBlob.hpp includes frame.hpp

Reply:
It's pretty hacky, but it might also work to add a #define _LOCORE and forward declare trapframe just before including sys/user.h.  I can't try this without arm hardware or maybe emulation, but that might work based on https://svnweb.freebsd.org/base/head/sys/arm/include/frame.h?revision=326258&view=markup 

Really the arm headers should do a similar thing to other arches and just forward declare what they need themselves rather than including frame.h, if for nothing other than consistency.

Reply:
This is caused by unwanted, unnecessary header pollution in FreeBSD should be handled in slightly more elegant way - by defining  #define _MACHINE_PCB_H_ before inclusion  of sys/user.h. See https://github.com/strejda/openjdk-jdk8u/commit/ab4be68dcf7e093482ee68208a19699e216fc693.
I will fix this problem for FreeBSD 14 (head) and 13 (stable), but I think that we can have this small workaround in openjdk  bsd port.